### PR TITLE
Adjust a few "since" values for 1.12.0

### PR DESCRIPTION
**Pull request by bonzini**
Add two "since 1.12.0" that were requested by @eli-schwartz:
* External programs as inputs and dependencies to custom targets, and external programs as dependencies to tests, were added as part of fixing #15774. #15774 is a regression where 1.10.x and earlier versions allowed overridden programs but 1.11 does not. I considered the bug to be that the result of `find_program()` was allowed in some cases and not in others, and would have considered that a bug worth fixing in 1.11.x. Eli disagreed when preparing 1.11.2, so bump the release introducing the fix from 1.11.2 to 1.12.0.
* mypy reported a potential AttributeError with Cython, which was fixed as part of completing the typing effort in 1.12.0. This later was reported as a genuine issue, caused by Generators that produce a non-`.pyx` source. Eli asks to consider this as a DSL extension and warn about it (which makes sense because non-Ninja backends don't support Cython at all). The situation is not handled by KwargInfo; detect it specifically as part of `add_target`'s postprocessing.

In both cases, any decision on whether to introduce the fixes in 1.11.3 can be taken later. These are an ugly corner case that sits between bug (incorrect typing, manifestly inconsistent DSL constraints) and new feature, and more discussion seems to be needed.

### CODE REVIEW

File: test cases/cython/2 generated sources/meson.build [hunk 1]
```diff
@@ -74,6 +75,26 @@ test(
...
+# source in the same target.
+gen_c = generator(
+  find_program('generator.py'),
+  arguments : ['@INPUT@', '@OUTPUT@'],
+  output : '@BASENAME@',
+)
+
+genc_ext = py3.extension_module(
+  'genc',
+  gen_c.process('genc.pyx.in', 'answer.c.in'),
```
Comment (bonzini) [inline test cases/cython/2 generated sources/meson.build:88]: `genc.pyx` is generated to work around #16001.

**Review by dcbaker — APPROVED** [review]
This looks good to me, but get Eli's thoughts too.

File: mesonbuild/interpreter/interpreter.py [hunk 2]
```diff
@@ -3501,6 +3501,17 @@ def add_target(self, name: str, tobj: build.Target) -> None:
...
             self.add_stdlib_info(tobj)
 
+            if tobj.uses_cython():
+                for gensrc in tobj.get_generated_sources():
+                    # a generator producing a .pyx has always worked; feeding
+                    # anything else from a generator into Cython triggered a bug
+                    # in the ninja backend and was effectively a new feature.
+                    if isinstance(gensrc, build.GeneratedList) and \
+                            any(not o.endswith('.pyx') for o in gensrc.get_outputs()):
+                        FeatureNew.single_use('Generators for Cython sources',
```
Comment (eli-schwartz) [inline mesonbuild/interpreter/interpreter.py]: Is it worth saying "for non-pyx Cython sources"?

**Review by eli-schwartz — APPROVED** [review]
Thanks. Overall this looks quite good.

File: mesonbuild/interpreter/interpreter.py (hunk 2 above)
Comment (bonzini) [reply mesonbuild/interpreter/interpreter.py]: Yes, absolutely.
